pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,476 @@
+}
+
+func (b *bitArray) isSet(index uint32) bool {
+	return (b.data[index/bitsUint32]>>(index%bitsUint32))&0x1 == 1
+}
+
+func (b *bitArray) findAndSetFirstZero() (uint32, error) {
+	for i := b.firstIndex; i < len(b.data); i++ {
+		if b.data[i] == 0xFFFFFFFF {
+			continue
+		}
+		for j := uint32(0); j < bitsUint32; j++ {
+			if (b.data[i]>>j)&0x1 == 0 {
+				v := uint32(i)*bitsUint32 + j
+				b.set(v)
+				b.firstIndex = int(i)
+				return v, nil
+			}
+		}
+	}
+	return 0, fmt.Errorf("could not find an empty slot")
+}
+
+func (b *bitArray) clear(index uint32) {
+	i := index / bitsUint32
+	if i < uint32(b.firstIndex) {
+		b.firstIndex = int(i)
+	}
+	b.data[i] &= ^(1 << (index % bitsUint32))
+}
+
+// length for the user namespace to create (65536).
+const userNsLength = math.MaxUint16 + 1
+
+// Limit the total number of pods using userns in this node to this value.
+// This is an alpha limitation that will probably be lifted later.
+const maxPods = 1024
+
+type usernsManager struct {
+	used     *bitArray
+	usedBy   map[string]uint32
+	removed  int
+	currPods int
+	kl       *Kubelet
+	sync.Mutex
+}
+
+// UserNamespace holds the configuration for the user namespace.
+type userNamespace struct {
+	// UIDs mappings for the user namespace.
+	UIDMappings []idMapping `json:"uidMappings"`
+	// GIDs mappings for the user namespace.
+	GIDMappings []idMapping `json:"gidMappings"`
+}
+
+// Pod user namespace mapping
+type idMapping struct {
+	// Required.
+	HostId uint32 `json:"hostId"`
+	// Required.
+	ContainerId uint32 `json:"containerId"`
+	// Required.
+	Length uint32 `json:"length"`
+}
+
+func MakeUserNsManager(kl *Kubelet) (*usernsManager, error) {
+	m := usernsManager{
+		// Create a bitArray for all the UID space (2^32).
+		// As a by product of that, no index param to bitArray can be out of bounds (index is uint32).
+		used:   makeBitArray((math.MaxUint32 + 1) / userNsLength),
+		usedBy: make(map[string]uint32),
+		kl:     kl,
+	}
+	// First block is reserved for the host.
+	m.used.set(0)
+
+	// Second block will be used for phase II. Don't assign that range for now.
+	m.used.set(1)
+
+	if kl != nil {
+		found, err := kl.listPodsFromDisk()
+		if err != nil {
+			return nil, fmt.Errorf("user namespace manager can't read pods from disk: %w", err)
+		}
+
+		for _, uid := range found {
+			if err := m.recordPodMappings(uid); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return &m, nil
+}
+
+// recordPodMappings registers the range used for the user namespace iif the
+// usernsConfFile exits in the pod directory.
+func (m *usernsManager) recordPodMappings(pod types.UID) error {
+	usernsConfFile := m.getUserNamespaceMappingsFile(pod)
+
+	// If the size is too big, something is fishy. Don't read it all in mem.
+	// Ignore errors, the file might not be created yet and afterwards we
+	// check which error was returned.
+	if info, err := os.Stat(usernsConfFile); err == nil && info.Size() >= 1024*1024 {
```

Comment:
I'd name the result of `1024*1024` into a variable, to make it clear what that number represents, and _then_ consume it.

Reply:
@sftim can you please elaborate why?

I don't see the need for a variable name. It is just an arbitrary low number that we won't ever exceed with the mappings, to make sure you can't make the kubelet use all the mem if you manage to be root and write a huge file. The kubelet won't make those kind of non-sense.

Reply:
The reason I'd name this const is to lower the barrier of entry to someone who wants to maintain or understand this code.
This is 1MiB, right? So maybe also comment that userns mappings files must be 1MiB or less and that anything larger is an error. The const could be `usernsMappingFileMaxSize`.

Reply:
I'm not sure I see the point. I think it is quite clear that a "big" size is being errored out, we even have a comment saying: `If the size is too big, something is fishy. Don't read it all in mem`. That clearly explains that we are avoiding to read the whole file in mem as big sizes seem fishy and are not expected. IMHO this is quite clear for a new comer why it is being done, because the comment just explains that.

Having a name for the 1MiB would not help to understand any of that, IMHO. The code is clear and has a comment explaining the intent too.

We can add a var for this, but really I don't think anything improves because of that.

Reply:
Ok, Giuseppe dropped this part. I don't really like it, but it seems probably true that if a malicious actor can write there, then we have bigger problems.